### DISCUSSION TITLE
Change "This Month" to display Month To Date view

### DISCUSSION THREAD

**Discussion by metmarkosaric** [plausible/analytics#391] [Anything Else]
"This Month" currently displays the full month which is not that useful, especially in the first half of the month which leaves most of the top graph empty and the comparison wrong too. If "This Month" simply shows month to date view rather than the full month would be much more useful. 

(request from [Twitter](https://twitter.com/nbgoodall/status/1322149369378938882))

**Comment by nbgoodall** [comment]
Thank you, Marko! 🚀

**Comment by ukutaht** [comment]
Agreed 💯

**Comment by Vigasaurus** [comment]
Am I misunderstanding or is this all that's wanted here? i.e. only the graph to change? Or do we want comparisons to also reference only the subset of the previous month?
![image](https://user-images.githubusercontent.com/19434157/102351837-b4f67280-3f6c-11eb-8693-b5f1f24efcc6.png)

**Reply by nbgoodall** [reply]
Ideally it'd update the comparisons too, like Stripe does. The question I'd like to answer is _"How does the month so far compare to the previous month?"_

**Reply by Vigasaurus** [reply]
Yeah that makes sense - I'll see what I can do 👍 we may need some sort of docs explanation for that, just to make it clear the difference between selecting "this month" vs "last 30 days"

**Comment by ukutaht** [comment]
A couple thoughts here @Vigasaurus:

I'm not sure if changing the graph is necessary. I get it if people want it but personally I quite like seeing the whole month/day and how far through we are. What do you think @metmarkosaric ? Seems like I'm in the minority here which is ok :)

The main point is getting the comparisons right. If I'm looking at the stats on the 16th of Dec, it should compare against 1-16 November.  I would change the name of the selection in the date dropdown to `Month to date` to make it clearer that's what it is.

There are edge cases where for example current month has 31 days and the previous has 29. This only kicks in at the end of the month. I haven't thought about how to solve this.

The comparison thing also applies to daily stats. If I'm looking at the stats for today at 3pm, it should compare with yesterday 12am-3pm, not the whole day.

**Reply by Vigasaurus** [reply]
> I'm not sure if changing the graph is necessary.

I personally don't think so either - I was able to come up with both a React and a direct-from-clickhouse way to do this, so if we wanted to it could be made a switch of sorts on the front-end. 

I think your points about the comparisons make a lot of sense, and they are about what I'm working towards - basically splitting off the previous-query generator for comparisons that can be partially complete and making their intervals manually from the first (or midnight) to the current query's time/date but minus one month/day. 
I figure to solve the whole different month lengths we just keep the difference - i.e. if I'm looking at March 1-30, I should compare to February 1-28 no matter what (and the day-based one will obviously be okay) - thoughts? 


Also, that does bring up a side question for me (I haven't really looked) - how do we handle daylight savings for timezones that have it?

**Reply by metmarkosaric** [reply]
I would personally prefer the graph to have the existing days/time only rather than be empty (this is especially a problem early on in a month where the biggest visual element of Plausible is pretty much empty) but it's not like this is a very critical request that we hear a lot about. So if it's much simpler to keep the graph as it is, it's probably not the biggest issue.

Comparison for sure is more important as right not it's not really useful. I would keep the comparison simpler though. Just compare it to exactly the same amount of time immediately prior to the time you selected in the date picker. So if I select Month To Date right now (December 1 to December 16), it should compare to the same amount of time ending on December 1st where the comparison starts (November 15th to November 30th). Then all these edge cases become not important. I think this is something that's more expected than that we actually compare to November 1 to November 16 which is not what I would personally expect it to be compared to.

The other comparison is more year over year or month over month which is normally used for different purposes in marketing rather than the "previous period" comparison. There's a request for that too here though: #277 

But I don't know if there's any difference in how simple or difficult it is to build between these two.

**Reply by Vigasaurus** [reply]
Hmm, that's an interesting take - your implementation would actually be much simpler to implement, but neither is particularly difficult. What I'm concerned about though is the potential for data not lining up how users would want in this model - i.e. I feel as though the same subset of a previous period is more intuitive and potentially more useful for all "to-date" periods (and your implementation is already the existing way that custom date ranges work, if a user wanted to use it like that).

**Reply by metmarkosaric** [reply]
thanks @Vigasaurus! 

i've been doing some research and can confirm that the way i described before is the default industry standard when people talk about comparisons. as it's also the simpler one to implement plus it's the way everything else works in Plausible, it makes sense to keep it that way for this feature too. 

in future, we may add month over month or year over year comparisons too, but just fixing this would be more than enough for majority of use cases.

so to confirm: if I select Month To Date right now (December 1 to December 16), it should compare to the same amount of time ending on December 1st where the comparison starts (November 15th to November 30th). thanks again for your help!

**Reply by nbgoodall** [reply]
> "i've been doing some research and can confirm that the way i described before is the default industry standard when people talk about comparisons."

My $0.02 as a customer…

I feel like you should be able to select the comparison period (which is how e.g. Stripe does it) as opposed to inferring it. E.g., how can I compare November to January? I realise that's another menu but would take out some of the ambiguity. Re. the above, I'm personally more interesting in the direct comparison to the previous month (Nov 1–16 vs Dec 1–16), but appreciate the other one might be more popular.

Thank you for the awesome product! :)

**Reply by metmarkosaric** [reply]
thanks @nbgoodall! yeah, that makes sense to have. this issue here is more to fix the current comparison which shows wrong data as it doesn't compare to the exact time period (if you select "this month" right now, it compares 2 weeks of december to the full month of november for instance rather than same amount of time in november). so that needs to be fixed as a short term thing but we do plan to add more comparisons in the future too.

**Reply by metmarkosaric** [reply]
So basically to confirm the two different things we'd like to do: 

One is that we'd like to fix the comparisons so they compare to the same exact amount of time. Currently they don't do that if the current time period is not completed so they are not so useful. So "this month" right now doesn't compare to 16 days of last month but to the full last month. same for "today", it doesn't compare to the same amount of time yesterday but to the full day yesterday. #344 

The second one is that we'd like to add few more comparisons in the future to make this even more useful such as year over year. There's a thread here for those ideas: #277 . I've included your ideas there too now @nbgoodall

**Reply by Vigasaurus** [reply]
So to confirm - this specific request would just need to properly slice comparisons that cut into the current month/day - 
such that instead of comparing Dec 1-Dec 16 to October 31-November 30, we'd only compare Dec 1-Dec 16 to Nov 14-Nov 30 (in the "this month" selection). 
and the same for "today" selection with the split being from midnight-current hour and yesterday this time yesterday-midnight. 

And then we'd leave the more accurate split on "same stretch" in the other feature with the better comparison choice?

**Reply by metmarkosaric** [reply]
Exactly. The comparison should compare to exactly the same amount of time in the previous period. This is something we don't do now and it affects many users so is of more immediate priority to fix and it will make Plausible more useful to many.

The better comparison choices we need to discuss in more detail as there are several different requests from people: we need to decide which of them to introduce (month over month, year over year, completely custom comparison to any period that you pick...), where to place them so they are easy to use/understand etc. Needs a bit more thinking before we know what to build.

**Reply by Vigasaurus** [reply]
👍 I'll work on the former for now then

**Comment by metmarkosaric** [comment]
"This Month" is now "Month to Date" with a comparison to the previous period
